File: app/MMarket/c/src/pgs/cosPg.c
// Copyright 19-Jan-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "pgs/cosPg.h"
#include "dmc/cgi.h"
#include "dmc/str.h"
#include "dmc/err.h"
#include "dmc/js.h"
#include "data/models.h"
#include "db/evalsDb.h"
#include "db/quotesTb.h"

char *cosPg_process (Mchar *mrq) {
  char *rq = cgi_rq_string(mrq, "rq");
  if (str_eq(rq, "idata")) {
    char *model_id = cgi_rq_string(mrq, "modelId");
    char *params_js = ochar_some(mchar_get(mrq, "params"));

    Model *md = models_get(model_id);
    AModelEval *evs = evalsDb_read(model_id)->evals;

    ADouble *params = NULL;
    ModelEval *eval = NULL;
    if (!js_is_null(params_js)) {
      params = aDouble_from_js(params_js);
      /**/int ffind (ModelEval *e) {
      /**/  return aDouble_eq(params, e->params, 0.000001);
      /**/}
      eval = oModelEval_nsome(aModelEval_find(evs, ffind));
      if (!eval) params_js = js_wn(); // Enter in next 'if'
    }
    if (js_is_null(params_js)) { // Warning: Not use 'else'
      /**/int fgreater (ModelEval *e1, ModelEval *e2) {
      /**/  return e1->hvalue < e2->hvalue;
      /**/}
      aModelEval_sort(evs, fgreater);
      eval = *(evs->es);
      params = eval->params;
    }

    Quotes *qs = quotesTb_read();
    Result *result = model_result(md, qs, params);

    Mchar *rp = mchar_new();
    mchar_put(rp, "model", model_to_js(md));
    mchar_put(rp, "result", result_to_js(result));
    mchar_put(rp, "eval", modelEval_to_js(eval));
    mchar_put(rp, "cos", achar_to_js(qs->cos));
    return cgi_rp(rp);
  } else if (str_eq(rq, "co")) {
    char *model_id = cgi_rq_string(mrq, "modelId");
    ADouble *params = aDouble_from_js(ochar_some(mchar_get(mrq, "params")));
    char *co = cgi_rq_string(mrq, "co");
    Model *md = models_get(model_id);
    Quotes *all_qs = quotesTb_read();
    int ico = quotes_ico(all_qs, co);
    if (ico == -1) FAIL (str_f("Company %s not found", co));
    ADouble *qs = quotes_closes(all_qs, ico);
    Quotes *co_qs = quotes_mk_single(all_qs, ico);
    Result *result = model_result(md, co_qs, params);
    ADouble *refs = model_refs(md, co_qs, params);

    Mchar *rp = mchar_new();
    mchar_put(rp, "result", result_to_js(result));
    mchar_put(rp, "dates", achar_to_js(all_qs->dates));
    mchar_put(rp, "qs", aDouble_to_js(qs));
    mchar_put(rp, "refs", aDouble_to_js(refs));
    return cgi_rp(rp);

  } else return FAIL(str_f("Unexpected value for 'rq': %s", rq));
}

File: app/KtMMarket/go/pgs/cosPg/cosPg.go
// Copyright 04-Jul-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// Company charts page.
package cosPg

import (
	"github.com/dedeme/KtMMarket/data/model"
	"github.com/dedeme/KtMMarket/data/modelEval"
	"github.com/dedeme/KtMMarket/data/order"
	"github.com/dedeme/KtMMarket/data/result"
	"github.com/dedeme/KtMMarket/db"
	"github.com/dedeme/KtMMarket/fns"
	"github.com/dedeme/ktlib/arr"
	"github.com/dedeme/ktlib/cgi"
	"github.com/dedeme/ktlib/js"
)

func Process(ck string, mrq cgi.T) string {
	rq := js.Rs(mrq["rq"])
	switch rq {
	case "idata":
		modelId := js.Rs(mrq["modelId"])
		paramsJs := mrq["params"]
		md := model.FromId(modelId)
		evs := db.EvalsDb(modelId).Read().Evals

		var params []float64
		var eval *modelEval.T
		if !js.IsNull(paramsJs) {
			params = arr.Map(js.Ra(paramsJs), js.Rd)
			var ok bool
			eval, ok = arr.Find(evs, func(ev *modelEval.T) bool {
				return fns.EqParams(ev.Params, params)
			})
			if !ok {
				paramsJs = js.Wn()
			}
		}
		if js.IsNull(paramsJs) { // Warning: Not use 'else'
			arr.Sort(evs, func(e1, e2 *modelEval.T) bool {
				return e1.Hvalue > e2.Hvalue
			})
			eval = evs[0]
			params = eval.Params
		}

		qs := db.QuotesTb().Read()
		orders, historic, _, _, _, _, _, profits := md.Simulation(qs, params)
		rs := result.New(
			historic[len(historic)-1],
			arr.Reduce(profits, 0, func(r, e float64) float64 {
				return r + e
			})/float64(len(profits)),
			float64(order.Sales(orders)),
		)

		return cgi.Rp(ck, cgi.T{
			"model":  model.ToJs(md),
			"result": result.ToJs(rs),
			"eval":   modelEval.ToJs(eval),
			"cos":    js.Wa(arr.Map(qs.Cos, js.Ws)),
		})

	case "co":
		modelId := js.Rs(mrq["modelId"])
		params := arr.Map(js.Ra(mrq["params"]), js.Rd)
		co := js.Rs(mrq["co"])
		md := model.FromId(modelId)
		allQs := db.QuotesTb().Read()
		coIx := allQs.CompanyIndex(co)
		if coIx == -1 {
			panic("Company " + co + " not found")
		}

		qs := allQs.GetSingle(coIx)
		orders, historic, _, _, refs, _, _, profits := md.Simulation(qs, params)
		rs := result.New(
			historic[len(historic)-1],
			arr.Reduce(profits, 0, func(r, e float64) float64 {
				return r + e
			})/float64(len(profits)),
			float64(order.Sales(orders)),
		)

		return cgi.Rp(ck, cgi.T{
			"result": result.ToJs(rs),
			"dates":  js.Wa(arr.Map(qs.Dates, js.Ws)),
			"qs": js.Wa(arr.Map(qs.Closes, func(c []float64) string {
				return js.Wd(c[0])
			})),
			"refs": js.Wa(arr.Map(refs, func(r []float64) string {
				return js.Wd(r[0])
			})),
		})

	default:
		panic("Value of rq (" + rq + ") is not valid")
	}
}
